File: common-auth/src/main/java/com/reservation/commonauth/auth/token/JwtTokenProvider.java
package com.reservation.commonauth.auth.token;

import java.util.Date;

import org.springframework.stereotype.Component;

import com.reservation.commonauth.auth.config.JwtProperties;

import io.jsonwebtoken.Claims;
import io.jsonwebtoken.ExpiredJwtException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.SignatureAlgorithm;
import lombok.RequiredArgsConstructor;

@Component
@RequiredArgsConstructor
public class JwtTokenProvider {
	private final JwtProperties jwtProperties;

	public String generateToken(Long userId, String role) {
		return createToken(userId, role, jwtProperties.getAccessTokenExpiry());
	}

	public String generateRefreshToken(Long userId, String role) {
		return createToken(userId, role, jwtProperties.getRefreshTokenExpiry());
	}

	private String createToken(Long userId, String role, long validityMillis) {
		Claims claims = Jwts.claims().setSubject(userId.toString());
		claims.put("role", role);

		Date now = new Date();
		Date expiry = new Date(now.getTime() + validityMillis);

		return Jwts.builder()
			.setClaims(claims)
			.setIssuedAt(now)
			.setExpiration(expiry)
			.signWith(SignatureAlgorithm.HS256, jwtProperties.getSecretKey())
			.compact();
	}

	public boolean isTokenValid(String token) {
		try {
			Jwts.parser()
				.setSigningKey(jwtProperties.getSecretKey())
				.parseClaimsJws(token);
			return true;
		} catch (ExpiredJwtException e) {
			throw e;
		} catch (Exception e) {
			return false;
		}
	}

	public Long extractUserId(String token) {
		try {
			return Long.parseLong(
				Jwts.parser()
					.setSigningKey(jwtProperties.getSecretKey())
					.parseClaimsJws(token)
					.getBody()
					.getSubject()
			);
		} catch (ExpiredJwtException e) {
			return Long.parseLong(e.getClaims().getSubject());
		} catch (Exception e) {
			return null;
		}
	}

	public String extractRole(String token) {
		return Jwts.parser()
			.setSigningKey(jwtProperties.getSecretKey())
			.parseClaimsJws(token)
			.getBody()
			.get("role", String.class);
	}

	public Date extractExpiration(String token) {
		try {
			return Jwts.parserBuilder()
				.setSigningKey(jwtProperties.getSecretKey())
				.build()
				.parseClaimsJws(token)
				.getBody()
				.getExpiration();
		} catch (ExpiredJwtException e) {
			return null;
		}
	}
}

File: identity-service/src/main/kotlin/com/msa/identityservice/auth/token/JwtTokenProvider.kt
package com.msa.identityservice.auth.token

import com.msa.identityservice.auth.token.dto.TokenAuthInfo
import com.msa.identityservice.auth.token.enums.Role
import com.msa.identityservice.config.properties.JwtProperties
import com.msa.identityservice.exception.BusinessErrorCode
import io.github.oshai.kotlinlogging.KotlinLogging
import io.jsonwebtoken.Claims
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.stereotype.Component
import java.nio.charset.StandardCharsets
import java.util.*


private val logger = KotlinLogging.logger {}

@Component
class JwtTokenProvider(
    private val jwtProperties: JwtProperties
) {

    fun generateJsonWebToken(
        jti: String,
        issuedAt: Date,
        expiration: Date,
        userId: Long,
        email: String,
        role: Role,
        deviceId: String
    ): String {
        val key = Keys.hmacShaKeyFor(jwtProperties.secretKey.toByteArray(StandardCharsets.UTF_8))

        return Jwts.builder()
            .header()
            .and()
            .id(jti)
            .issuer("identity-service")
            .issuedAt(issuedAt)
            .expiration(expiration)
            .subject(userId.toString())
            .claim("role", role.name)
            .claim("email", email)
            .claim("deviceId", deviceId)
            .signWith(key)
            .compact()
    }

    fun getClaims(token: String): Claims {
        val key = Keys.hmacShaKeyFor(jwtProperties.secretKey.toByteArray(StandardCharsets.UTF_8))

        return Jwts.parser()
            .verifyWith(key)
            .build()
            .parseSignedClaims(token)
            .payload
    }

    fun extractAuthInfo(token: String): TokenAuthInfo {
        try {
            val claims = getClaims(token)

            return extractAuthInfo(claims)
        } catch (e: Exception) {
            logger.error { "Extract AuthInfo exception (token): $e" }
            throw BusinessErrorCode.UNAUTHORIZED.exception("인증 정보가 올바르지 않습니다.")
        }
    }

    fun extractAuthInfo(claims: Claims): TokenAuthInfo {
        try {
            val jti = claims.id
            val userId = claims.subject.toLong()
            val role = claims.get("role", String::class.java)
            val email = claims.get("email", String::class.java)
            val expiration = claims.expiration
            val deviceId = claims.get("deviceId", String::class.java)

            return TokenAuthInfo(
                jti = jti,
                userId = userId,
                email = email,
                role = Role.valueOf(role),
                expiresAt = expiration,
                deviceId = deviceId
            )
        } catch (e: Exception) {
            logger.error { "Extract AuthInfo exception (Claims): $e" }
            throw BusinessErrorCode.UNAUTHORIZED.exception("인증 정보가 올바르지 않습니다.")
        }
    }

}
